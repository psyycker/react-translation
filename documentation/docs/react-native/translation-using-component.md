---
sidebar_position: 1
---

# Translation using component (React Native)

The translation using JSX component is similar in React Native than in React.
For maintenance reasons, React and React Native are using different components and the import for React Native
needs to be done using Tree Shaking (See example)

```jsx
import TranslatedText from "@psyycker/react-translation/build/react-native/translated-text";

function MyComponent() {

  return <div><TranslatedText translationKey="myKey" defaultValue="my default value"/></div>
}
```
