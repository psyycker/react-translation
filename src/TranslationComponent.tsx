import * as React from "react";
import {useTranslation} from "./hooks";

export type TranslationComponentProps = {
  translationKey: string;
  defaultValue: string;
  parameters?: object;
  style?: any
}

let rightFunction: React.ReactElement<TranslationComponentProps>;

export function generateRNComponent(Text: any) {
  return function TranslationComponentRN({translationKey, defaultValue, parameters, style}: TranslationComponentProps): React.ReactElement<TranslationComponentProps> {
    const { getTranslation } = useTranslation();
    return (
      <Text style={style}>
        {getTranslation({
          translationKey, defaultValue, parameters
        })}
      </Text>
    )
  }
}

export function TranslationComponentWeb({translationKey, defaultValue, parameters}: TranslationComponentProps): React.ReactElement<TranslationComponentProps> {
  const { getTranslation } = useTranslation();
  return (
    <>
      {getTranslation({
        translationKey, defaultValue, parameters
      })}
    </>
  )
}

try {
  const { Text } = require("react-native");
  // @ts-ignore
  rightFunction = generateRNComponent(Text);
} catch (e) {
  // @ts-ignore
  rightFunction = TranslationComponentWeb;
}

export default rightFunction;
