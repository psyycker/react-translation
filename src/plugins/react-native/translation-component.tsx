import React from "react"
import {Text} from "react-native";
import useTranslation from "../../hooks/use-translation";

export type TranslationComponentProps = {
  translationKey: string;
  defaultValue: string;
  parameters?: object;
  style?: any
}

const TranslationComponent =({translationKey, defaultValue, parameters, style}: TranslationComponentProps): React.ReactElement<TranslationComponentProps> => {
  const { getTranslation } = useTranslation();
  return (
    <Text style={style}>
      {getTranslation({
        translationKey, defaultValue, parameters
      })}
    </Text>
  )
}

export default TranslationComponent
