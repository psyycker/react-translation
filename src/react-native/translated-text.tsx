/* istanbul ignore file */
// No way to test this as it's react native and the project is react based

import React from 'react';
/* eslint-disable */
// @ts-ignore
import { Text } from 'react-native';
import useTranslation from '../hooks/use-translation';
import { TranslationComponentProps } from '../types';

function TranslatedText({
  translationKey, defaultValue, parameters, style,
}: TranslationComponentProps): React.ReactElement<TranslationComponentProps> {
  const { getTranslation } = useTranslation();
  return (
    <Text style={style}>
      {getTranslation({
        translationKey, defaultValue, parameters,
      })}
    </Text>
  );
}

export default TranslatedText;
