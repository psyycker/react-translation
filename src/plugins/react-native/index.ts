import TranslationComponent from "./translation-component";

export default function apply(reactTranslation: any) {
  console.log(reactTranslation)
  reactTranslation.Translation = TranslationComponent
}
