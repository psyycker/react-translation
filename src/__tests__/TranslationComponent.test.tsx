import * as React from "react";
import { render as renderReact } from "@testing-library/react";
import { render as renderRN } from "@testing-library/react-native";
import {generateRNComponent, TranslationComponentWeb} from "../TranslationComponent";
import { Text } from "react-native";
import {registerTranslations, setTranslationConfig} from "../index";

const TranslationComponentRN = generateRNComponent(Text);

describe('Translation component', function () {

  beforeAll(() => {
    setTranslationConfig({
      defaultLocale: "en-US"
    })
    registerTranslations({
      "en-US": {
        "component": {
          "title": "myTranslation"
        }
      }
    });
  })

  describe('React', function () {
    it("Should do the translation", () => {
      const { asFragment, findByText } = renderReact(<TranslationComponentWeb translationKey="component.title" defaultValue="test" />)
      expect(asFragment()).toMatchSnapshot();
      expect(() => findByText("myTranslation")).not.toThrow();
    })
  });

  // Find a solution to that
  describe.skip('React Native', function () {
    it("Should do the translation", () => {
      const { findByText } = renderRN(<TranslationComponentRN translationKey="component.title" defaultValue="test" />)
      expect(() => findByText("myTranslation")).not.toThrow();
    })
  });
});
