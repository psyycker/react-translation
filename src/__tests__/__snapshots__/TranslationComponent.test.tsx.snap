// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation component Should do the translation 1`] = `
<DocumentFragment>
  value
</DocumentFragment>
`;
